### PR TITLE
[Resolve #1252] Supporting resolvers in Hook and Resolver arguments, with new !substitute, !join, !split, and !select resolvers!

### LINKED ISSUE #1252: Question: Join Resolver output with additional strings possible?
I'm wondering how you can use resolver outputs with some additional strings - e.g. use the stack_output and use some additional strings to attach / join them with. Can Jinja2 be use for this?

Usecase for example is to build a connection string (host:hostname:port?option=A), e.g.

parameters:
  paramA: `host:!stack_output <stack_name>::output:1234?option=A`

Dose the new stack_attr help here, e.g.

host_endpoint: !stack_output <stack_name>::output

parameters:
  paramA: `host:{{ stack_attr.host_endpoint }}:1234?option=A`

### PR DESCRIPTION
**Pull request by jfalkenstein**
This pull request brings the ability for Hooks and Resolvers to have resolvers in their arguments and have those resolvers automatically resolve when `self.argument` is accessed. When a resolver or hook is `setup`, it will also setup all nested resolvers in its argument. And when it is attached to a Stack, it and its argument will be recursively cloned so that no resolver is associated with a different stack. This will let resolvers be "inherited" from parent stack configurations.

## PR Checklist

- [x] Wrote a good commit message & description [see guide below].
- [x] Commit message starts with `[Resolve #issue-number]`.
- [x] Added/Updated unit tests.
- [x] Added/Updated integration tests (if applicable).
- [x] All unit tests (`make test`) are passing.
- [x] Used the same coding conventions as the rest of the project.
- [x] The new code passes pre-commit validations (`pre-commit run --all-files`).
- [x] The PR relates to _only_ one subject with a clear title.
      and description in grammatically correct, complete sentences.

## Approver/Reviewer Checklist

- [ ] Before merge squash related commits.

## Other Information

[Guide to writing a good commit](http://chris.beams.io/posts/git-commit/)

### CODE REVIEW

File: sceptre/hooks/__init__.py [hunk 1]
```diff
@@ -20,21 +22,12 @@ class Hook(abc.ABC):
...
-        self.argument = argument
-        self.stack = stack
-
-    def setup(self):
-        """
-        setup is a method that may be overwritten by inheriting classes. Allows
-        hooks to run so initalisation steps when config is first read.
-        """
-        pass  # pragma: no cover
-
```
Comment (jfalkenstein) [inline sceptre/hooks/__init__.py:37]: This is now defined in the base class

File: sceptre/hooks/__init__.py [hunk 2]
```diff
@@ -43,15 +36,6 @@ def run(self):
...
 
-    def clone(self, stack: "Stack") -> "Hook":
-        """
-        Produces a "fresh" copy of the Hook, with the specified stack.
-
-        :param stack: The stack to set on the cloned resolver
-        """
-        clone = type(self)(self.argument, stack)
-        return clone
-
```
Comment (jfalkenstein) [inline sceptre/hooks/__init__.py:54]: This is now defined on the base class

File: sceptre/hooks/__init__.py [hunk 3]
```diff
@@ -1,39 +1,21 @@
...
-from typing import TYPE_CHECKING, Any
+from sceptre.resolvers import CustomYamlTagBase
 
 if TYPE_CHECKING:
     from sceptre.stack import Stack
 
 
-class Hook(abc.ABC):
+class Hook(CustomYamlTagBase, metaclass=abc.ABCMeta):
```
Comment (jfalkenstein) [inline sceptre/hooks/__init__.py:13]: The big change here is that, since there's very little that differentiates a hook from a resolver, other than the "resolve" vs "run" public method, they now share a common base class. A lot of the functionality on Hooks has now been moved up into that base class since it was shared with Resolvers.

File: sceptre/hooks/__init__.py [hunk 4]
```diff
@@ -84,7 +57,7 @@ def __set__(self, instance: "Stack", value):
         """
 
         def setup(attr, key, value: Hook):
-            attr[key] = clone = value.clone(instance)
+            attr[key] = clone = value.clone_for_stack(instance)
```
Comment (jfalkenstein) [inline sceptre/hooks/__init__.py:60]: `clone_for_stack` (as you'll see below) is a recursive clone operation that allows for cloning resolvers inside of the arguments.

File: sceptre/resolvers/__init__.py [hunk 5]
```diff
@@ -17,37 +17,158 @@
...
         initial setup of the object.
         """
         pass  # pragma: no cover
 
+    def __str__(self):
+        as_str = f"!{self.__class__.__name__}"
+        if self._argument is not None:
+            as_str += f"({self._argument})"
+
+        return as_str
```
Comment (jfalkenstein) [inline sceptre/resolvers/__init__.py:164]: This was mostly moved from the placeholders.py file. Previously, placeholders would access `resolver.argument` to represent the resolver when it couldn't be resolved. But if we're going to allow nested resolvers, that won't work. Instead, we need to represent the raw resolvers _all the way down_. Defining a `__str__` method on this base class and using the stringification of `self._argument` ensures that if any resolver in the argument cannot be resolved, it's placeholders end up being used all the way down into its arguments.

File: sceptre/resolvers/__init__.py [hunk 6]
```diff
@@ -17,37 +17,169 @@
...
+        be converted to useful placeholder values.
+        """
+        as_str = f"!{self.__class__.__name__}"
+        if self._argument is not None:
+            as_str += f"({self._argument})"
+
+        return as_str
+
+
+class Resolver(CustomYamlTagBase, metaclass=abc.ABCMeta):
```
Comment (jfalkenstein) [inline sceptre/resolvers/__init__.py:178]: I've moved most of the functionality that _was_ in the Resolver class into the CustomYamlTagBase, which is now shared by both the Resolver and Hook classes.

File: sceptre/resolvers/__init__.py [hunk 7]
```diff
@@ -127,23 +251,6 @@ def _no_recursive_get(self, stack: "stack.Stack"):
...
-        :param resolver: The Resolver to clone and set up
-        :return: The cloned resolver.
-        """
-        # We clone the resolver when we assign the value so that every stack gets its own resolver
-        # rather than potentially having one resolver instance shared in memory across multiple
-        # stacks.
-        clone = resolver.clone(stack)
-        clone.setup()
-        return clone
-
```
Comment (jfalkenstein) [inline sceptre/resolvers/__init__.py:146]: This method is no longer necessary. It's all been moved into the CustomYamlTagBase's `clone_for_stack` method.

File: sceptre/helpers.py [hunk 8]
```diff
@@ -68,6 +68,28 @@ def func_on_instance(key):
     return attr
 
 
+Container = Union[list, dict]
+Key = Union[str, int]
+
+
+def delete_keys_from_containers(keys_to_delete: List[Tuple[Container, Key]]):
```
Comment (jfalkenstein) [inline sceptre/helpers.py:75]: This is mostly just extracted out of the ResolvableProperty class and put here. I needed to also use this in the CustomYamlTagBase class.

File: sceptre/resolvers/placeholders.py [hunk 9]
```diff
@@ -65,10 +65,7 @@ def _create_explicit_resolver_placeholder(resolver: "resolvers.Resolver") -> str
     :param resolver: The resolver to create a placeholder for
     :return: The placeholder value
     """
-    base = f"!{type(resolver).__name__}"
-    suffix = f"({resolver.argument})" if resolver.argument is not None else ""
-    # double-braces in an f-string is just an escaped single brace
-    return f"{{ {base}{suffix} }}"
```
Comment (jfalkenstein) [inline sceptre/resolvers/placeholders.py:71]: This functionality has been moved to the CustomYamlTagBase class's `__repr__` method, where it accounts for arguments properly.

File: docs/_source/docs/resolvers.rst [hunk 10]
```diff
@@ -81,6 +105,49 @@ Refer to `sceptre-resolver-cmd <https://github.com/Sceptre/sceptre-resolver-cmd/
 
 .. _stack_attr_resolver:
 
+select
+~~~~~~
+
+This resolver allows you to select a specific index of a list of items. This is great for combining
+with the ``!split`` resolver to obtain part of a string. This function works almost the same as
+CloudFormation's ``!Select`` intrinsic function, **except you can use this with negative indexes to
+select with a reverse index**.
```
Comment (dboitnot) [inline docs/_source/docs/resolvers.rst:114]: What about "except you can use this with negative indices to select from the end of the list." ?

File: docs/_source/docs/resolvers.rst [hunk 11]
```diff
@@ -181,6 +248,32 @@ Example:
    parameters:
      VpcIdParameter: !stack_output_external prj-network-vpc::VpcIdOutput prod
 
+
+substitute
```
Comment (dboitnot) [inline docs/_source/docs/resolvers.rst:252]: I don't have super-strong feelings about this, but why "substitute" and not "sub" as in CF? I might argue in favor of "sub" for consistency with CF.

File: sceptre/helpers.py (hunk 8 above)
Comment (dboitnot) [reply sceptre/helpers.py:75]: I'm glad that this has been extracted but I'm a bit queasy about this function mutating collections it doesn't own. I got side-tracked by work but I feel this could be simplified (in the future) using list comprehensions and/or `filter`. But then I wasn't able to dig enough into how it's used to propose a refactor.

**Review by dboitnot — COMMENTED** [review]
I think this is great and will be a boon to users!

Looking at how resolvers work, though, I'm wondering if a more functional approach might drastically improve the code. There's a lot of mutation and behavior that, while correct, is not obvious and might be harder to reason about or troubleshoot.

It feels like resolvers (and maybe hooks?) could be evaluated as a series of transformations that act on data and their invocation is basically an abstract syntax tree.

The correct reply to the above is, of course, "Put your PR where your mouth is Dan." :)

File: docs/_source/docs/resolvers.rst (hunk 10 above)
Comment (jfalkenstein) [inline docs/_source/docs/resolvers.rst:114]:
```suggestion
CloudFormation's ``!Select`` intrinsic function, **except you can use this with negative indices to
select from the end of a list**.
```

File: docs/_source/docs/resolvers.rst (hunk 11 above)
Comment (jfalkenstein) [reply docs/_source/docs/resolvers.rst:252]: I don't know. I guess I just don't like the abbreviation. It doesn't seem as clear as `substitute`. @zaro0508 , thoughts on this?

File: sceptre/helpers.py (hunk 8 above)
Comment (jfalkenstein) [reply sceptre/helpers.py:75]: A couple responses:
1. It _does_ own them. That is why there is the cloning functionality of resolvers. The only collections that are mutated are ones that are effectively deep-copied when they are assigned to the stack.
2. This is generally the way resolvers have always worked. The basic idea is that it prevents us from ever needing to resolve the same resolver twice for the same stack. This can be important, since the resolvable attribute might be accessed a _number_ of times over the course of execution.
3. Once we've resolved the resolver, there's no sense in keeping it around; its purpose has been fulfilled.

**Comment by jfalkenstein** [comment]
On the subject of in-place mutation...
1. Because of the cloning employed when a resolver is associated with a stack, it effectively owns the data at that point. The only collections that are mutated are ones that are effectively deep-copied when they are assigned to the stack.
2. This is generally the way resolvers have always worked. The basic idea is that it prevents us from ever needing to resolve the same resolver twice for the same stack. This can be important, since the resolvable attribute might be accessed a _number_ of times over the course of execution.
3. Once we've resolved the resolver, there's no sense in keeping it around; its purpose has been fulfilled.

File: docs/_source/docs/resolvers.rst (hunk 11 above)
Comment (dboitnot) [reply docs/_source/docs/resolvers.rst:252]: I agree "substitute" has more intrinsic clarity, but since most of our users are going to be pretty familiar with CF and `sub` I think we could avoid the mental dichotomy by sticking to CF's nomenclature. But I would like @zaro0508's thoughts as well.

**Review by dboitnot — APPROVED** [review]

File: docs/_source/docs/resolvers.rst (hunk 11 above)
Comment (jfalkenstein) [reply docs/_source/docs/resolvers.rst]: I've changed this to `sub`

**Comment by quinnjr** [comment]
I'm really looking forward to this PR being merged

File: tests/test_resolvers/test_sub.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+class TestSub:
+    def test_resolve__combines_resolvers_into_single_string(self):
+        argument = [
+            "{first} is {first_value}; {second} is {second_value}",
+            {
+                "first": FirstResolver(),
+                "second": SecondResolver(),
+                "first_value": 123,
+                "second_value": 456,
+            },
```
Comment (zaro0508) [inline tests/test_resolvers/test_sub.py:26]: how about a test when one of the resolvers returns a non-string (list or dict) object?

how about a test with a missing replacement parameter? 
```
        argument = [
            "{first} and {second}",
            {
                "second": "B"
            },
``` 

How about a test where resolver returns `None`?
```
        argument = [
            "{first} and {second}",
            {
                "first": ResolverReturnsNone(),
                "second": "B"
            },
```

File: tests/test_resolvers/test_join.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    def resolve(self):
+        return "first"
+
+
+class SecondResolver(Resolver):
+    def resolve(self):
+        return "last"
+
+
+class TestJoin:
```
Comment (zaro0508) [inline tests/test_resolvers/test_join.py:17]: would it make sense to test when a resolver returns non-string values to join? also a test when resolver returns None?

File: setup.py [hunk 14]
```diff
@@ -71,7 +71,11 @@ def get_version(rel_path):
             "stack_output_external ="
             "sceptre.resolvers.stack_output:StackOutputExternal",
             "no_value = sceptre.resolvers.no_value:NoValue",
+            "select = sceptre.resolvers.select:Select",
             "stack_attr = sceptre.resolvers.stack_attr:StackAttr",
+            "sub = sceptre.resolvers.substitute:Sub",
```
Comment (zaro0508) [inline setup.py:76]: shouldn't this be `sub = sceptre.resolvers.sub:Sub`?

**Review by zaro0508 — CHANGES_REQUESTED** [review]

File: docs/_source/docs/resolvers.rst [hunk 15]
```diff
@@ -81,6 +105,49 @@ Refer to `sceptre-resolver-cmd <https://github.com/Sceptre/sceptre-resolver-cmd/
...
+The argument for this resolver should be a list with two elements: (1) The delimiter to split on and
+(2) a string to split.
+
+Example:
+
+.. code-block:: yaml
+
+   notifications: !split
+     - ";"
+     - !stack_output my/sns/topics.yaml::SemicolonDelimitedArns
```
[the remark this answers is not in the document]
Comment (jfalkenstein) [reply docs/_source/docs/resolvers.rst:148]: Actually, I don't really. !sub, !join, !select, and !split are ONLY useful when you have resolvers to string together; They're clumsy if you don't. It's better to use Jinja to do these things, if you can. But if you have resolvers whose resolved values you need to manipulate/interpolate/select from, they're really the only way to do that. 

As such, If we don't show how these resolvers can be used with other resolvers, we're not really showing how to use them well.

File: docs/_source/docs/resolvers.rst [hunk 16]
```diff
@@ -181,6 +248,32 @@ Example:
...
+curly-brace templates to indicate variables, and (2) a dictionary where the keys are the format
+string's variable names and the values are the variable values.
+
+Example:
+
+.. code-block:: yaml
+
+   parameters:
+     ConnectionString: !sub
+       - "postgres://{username}:{password}@{hostname}:{port}/{database}"
```
[the remark this answers is not in the document]
Comment (jfalkenstein) [reply docs/_source/docs/resolvers.rst:269]: `!sub` _needs_ a second argument that defines the variable values. It's not the same as CloudFormation's sub that lets you just resolve values without defining them. It's just a glorified `str.format()` invocation. It doesn't have a "scope" outside its own arguments. If you want to resolve something from the Stack scope, you could combine `!sub` and Jinja syntax together, like this:

```python
!sub
  # Note the double-curly braces is JINJA syntax to access variables within Jinja scope
  - "postgres://{{ var.username }}:{password}@{hostname}:{port}/{{ database }}"
  - password: !ssm /my/ssm/password
     hostname: !stack_output my/database/stack.yaml::HostName
     port: !stack_output my/database/stack.yaml::Port
```

Notice how, if it's accessible within Jinja, you don't actually need to provide that to `!sub` as a variable, because Jinja interpolation will happen BEFORE the yaml is rendered. I'm going to look into clarifying this in the docs, since I doubt you'll be the first one confused about this.

File: docs/_source/docs/stack_config.rst [hunk 17]
```diff
@@ -99,7 +99,7 @@ and that Stack need not be added as an explicit dependency.
 
 hooks
 ~~~~~
-* Resolvable: No
+* Resolvable: No (but you can use resolvers _in_ hook arguments!)
```
[the remark this answers is not in the document]
Comment (jfalkenstein) [reply docs/_source/docs/stack_config.rst:102]: Yeah, I'll look into that.

File: sceptre/resolvers/placeholders.py [hunk 18]
```diff
@@ -65,10 +65,7 @@ def _create_explicit_resolver_placeholder(resolver: "resolvers.Resolver") -> str
...
+    return f"{{ {resolver} }}"
```
[the remark this answers is not in the document]
Comment (jfalkenstein) [reply sceptre/resolvers/placeholders.py:68]: Yes! Notice how it puts the resolver's repr in curly braces. It's actually doing something. (The double-curly braces are escaped curly braces in f-strings).

File: setup.py (hunk 14 above)
Comment (jfalkenstein) [reply setup.py:76]: Yikes! Yes! That's probably why my tests are failing.

File: tests/test_resolvers/test_join.py (hunk 13 above)
Comment (jfalkenstein) [reply tests/test_resolvers/test_join.py:17]: Great suggestions. I suppose in the first case, I'll want it to coerce the values to strings. In the second case, they'll just be removed from the list. I'll add those tests!

Comment (jfalkenstein) [reply tests/test_resolvers/test_join.py:12]: Done!

File: tests/test_resolvers/test_select.py [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+    def resolve(self):
+        return ["first", "second", "third"]
+
+
+class ItemResolver(Resolver):
+    def resolve(self):
+        return self.argument
+
+
+class TestSplit:
```
[the remark this answers is not in the document]
Comment (jfalkenstein) [reply tests/test_resolvers/test_select.py:17]: Yes!

File: tests/test_resolvers/test_split.py [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+
+
+class MyResolver(Resolver):
+    def resolve(self):
+        return "first,second,third"
+
+
+class TestSplit:
+    def test_resolve__splits_resolver_value_into_list(self):
+        argument = [",", MyResolver()]
```
[the remark this answers is not in the document]
Comment (jfalkenstein) [reply tests/test_resolvers/test_split.py:14]: Well, if it's not a string, it'll just blow up. However... it's hard to imagine a scenario where somebody would want to split a value that isn't a string. What sort of behavior would you expect for non-string arguments?

File: tests/test_resolvers/test_select.py [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+    def resolve(self):
+        return ["first", "second", "third"]
+
+
+class ItemResolver(Resolver):
+    def resolve(self):
+        return self.argument
+
+
+class TestSelect:
```
Comment (jfalkenstein) [inline tests/test_resolvers/test_select.py:17]: I need to implement better error handling around missing indexes/keys.

File: docs/_source/docs/resolvers.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (dboitnot) [reply docs/_source/docs/resolvers.rst:146]: ¿Por qué no los dos? :)

```yaml
   notifications: !split
     - ";"
     - "note1; note2; note3"
```

Resolves to:

```yaml
   notifications: [ "note1", "note2", "note3" ]
```

As a more practical example, the following will split a semicolon-delimited list exported by another stack:

```yaml
   notifications: !split
     - ";"
     - !stack_output my/sns/topics.yaml::SemicolonDelimitedArns
```

File: tests/test_resolvers/test_select.py (hunk 21 above)
Comment (dboitnot) [reply tests/test_resolvers/test_select.py:17]: Pydantic has changed my life:

https://docs.pydantic.dev/

File: tests/test_resolvers/test_split.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (zaro0508) [reply tests/test_resolvers/test_split.py:14]: i'm thinking in a case where the resolver returned something that was unexpected, like a list instead of a string?

File: docs/_source/docs/resolvers.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (jfalkenstein) [reply docs/_source/docs/resolvers.rst:146]: @dboitnot, sure, that _works_. But If you wanted to split a simple string, you'd just go like this:

```yaml
notifications: {{ "note1; note2; note3".split(";") }}
```

I fear that if we show people examples of using these resolvers in cases where you don't actually _need_ to use them (and where you really _shouldn't_ use them), it will be more confusing. The only times you need to use these resolvers is when you need to combine/manipulate the outputs of other resolvers. I actually added documentation to that effect to make it absolutely clear.

File: docs/_source/docs/stack_config.rst (hunk 17 above)
[the remark this answers is not in the document]
Comment (jfalkenstein) [reply docs/_source/docs/stack_config.rst:102]: Docs have been added.

File: tests/test_resolvers/test_split.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (jfalkenstein) [reply tests/test_resolvers/test_split.py:17]: Additional validation and tests have been added.

File: tests/test_resolvers/test_sub.py (hunk 12 above)
Comment (jfalkenstein) [reply tests/test_resolvers/test_sub.py:29]: Additional validation and tests have been added.

**Review by zaro0508 — CHANGES_REQUESTED** [review]
generally LGTM.  just a few minor fixes and should be good to go.

File: docs/_source/docs/hooks.rst [hunk 22]
```diff
@@ -224,7 +226,7 @@ Assume a Sceptre `copy` hook that calls the `cp command`_:
        - !copy
            options: "-r"
            source: "from_dir"
-           destination: "to_dir"
+           destination: !stack_output my/other/stack::CopyDestination
```
[the remark this answers is not in the document]
Comment (jfalkenstein) [reply docs/_source/docs/hooks.rst:229]: No. `!copy` in this imaginary example takes a dict as an argument, not a list. However, `after_update` and `before_update` should be lists. So it's correct in the docs, not in your suggestion.

**Review by zaro0508 — APPROVED** [review]
